### PR TITLE
CP-11136 Virtual Hardware Platform Version (master)

### PR DESCRIPTION
**Pull request by thomassa**
This change requires changes in other repositories:

* guest-templates needs to supply a value (initially `0L` by default) for the VM virtual hardware platform version when creating a VM. (This needs to be done immediately after the change to xen-api. I have a branch with this change, ready to push after this pull-request is merged. There is a mirror at https://github.com/xenserver/guest-templates/)

* xen-api-sdk needs to cope with fields of type `Set(Int)`. That should be done before xen-api is changed, so I shall leave this pull-request closed until that has been done. (Done now.)

This change is essentially a forward-port of #2169.

### CODE REVIEW

**Comment by thomassa** [comment]
xen-api-sdk can handle Set(Int) now, so I am re-opening this pullreq.

**Comment by thomassa** [comment]
(Not building at present: `Error: Unbound module Client.Client` in ocaml/xapi/xapi_vm_helper.ml)

**Comment by cheng--zhang** [comment]
Hi @thomassa 
When I trying to include those change to trunk-ring3 and try to build XAPI, seems it failed with: 

File "xapi_vm_helpers.ml", line 277, characters 5-62:
Error: Unbound module Client.Client

**Comment by djs55** [comment]
We normally "open Client" at the top, and then use "Client.VM.foo" rather
than "Client.Client.VM.foo".

The exception to this is when we have multiple Clients-- for example a
xenops client and a XenAPI client. In that case I recommend saying "module
XenAPI = Client.Client" at the top.

On Tue, Mar 17, 2015 at 8:47 AM, Cheng Zhang <notifications@github.com>
wrote:

> Hi @thomassa <https://github.com/thomassa>
> When I trying to include those change to trunk-ring3 and try to build
> XAPI, seems it failed with:
>
> File "xapi_vm_helpers.ml", line 277, characters 5-62:
> Error: Unbound module Client.Client
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xapi-project/xen-api/pull/2180#issuecomment-82205951>.
>



-- 
Dave Scott

**Comment by thomassa** [comment]
@xen-git Retest this please

Fixed by removing `open Xenstore` because `Client` was binding to `Xenstore.Client` rather than to the main `Client` module.

It turns out that we aren't using anything from Xenstore anyway.

@djs55, do you think it's better to use `open Client` rather than `Client.Client` even though we're using the module in just one place? (`Client.Client` keeps the file shorter by an `open ` and makes obvious what it is from `Client` that we are using.)

File: ocaml/idl/datamodel.ml [hunk 1]
```diff
@@ -64,7 +64,7 @@ let creedence_release_schema_major_vsn = 5
 let creedence_release_schema_minor_vsn = 72
 
 let cream_release_schema_major_vsn = 5
-let cream_release_schema_minor_vsn = 72
+let cream_release_schema_minor_vsn = 73
 
 let dundee_release_schema_major_vsn = 5
 let dundee_release_schema_minor_vsn = 83
```
Comment (thomassa) [inline ocaml/idl/datamodel.ml]: Should we set `dundee_release_schema_minor_vsn` equal to (or defined as) `schema_minor_vsn` or should that wait for now?

**Comment by djs55** [comment]
@thomassa now that we have so many clients (in the beginning there was only 1!) we should probably make a convention. I'd quite like to use `XenAPI.VM.start` and `Xenops.VM.start` or something similar.

**Comment by thomassa** [comment]
@djs55 Sounds reasonable. I'll do that.

**Comment by thomassa** [comment]
Done.

**Comment by cheng--zhang** [comment]
@thomassa can we merge pull request #2197 and merge the code back to creedence-sp branch? Which means close this pull request but use #2197 instead.

**Comment by robhoes** [comment]
I discussed this with Thomas, and we concluded that it is best to get this pull request merged into the master branch sooner rather than later, because it is essentially a cherry-pick of the Virtual Hardware Platform Version patches from creedence-bugfix and we want master to be up-to-date with what's in there. It would be clearer to separate this from any follow-up patches for the Windows Update stuff.

**Comment by robhoes** [comment]
I was about to suggest that some of these patches should be squashed together, because some of the later ones fix the earlier ones. But then I noticed that the first four commits are cherry-picked from creedence-bugfix, and the following ones make it work on master... I guess this is OK, because it makes it easy to compare the two branches.

**Comment by robhoes** [comment]
Since this is mostly a forward port plus some minor fixes (which look fine), I'll merge this now.

**Pull request merged**